offchain/advance-runner/src/snapshot/fs_manager.rs

```diff
@@ -213,26 +221,26 @@ impl SnapshotManager for FSSnapshotManager {
         Ok(())
     }
 
-    /// Reads the binary contents of the hash file in snapshot's directory
-    /// and converts them to a `Hash`. It assumes the file was created correctly
-    /// and makes no checks in this regard.
     #[tracing::instrument(level = "trace", skip_all)]
-    async fn get_template_hash(
-        &self,
-        snapshot: &Snapshot,
-    ) -> Result<Hash, FSSnapshotError> {
-        let path = snapshot.path.join(HASH_FILE);
-        tracing::trace!(?path, "opening hash file at");
-        let file = File::open(path.clone())
-            .context(OpenHashSnafu { path: path.clone() })?;
+    async fn validate(&self, snapshot: &Snapshot) -> Result<(), Self::Error> {
+        if self.config.validation_enabled {
+            let offchain_hash = snapshot.get_hash().await?;
+
+            let onchain_hash = crate::dapp_contract::get_template_hash(
+                self.config.dapp_address.clone(),
+                self.config.provider_http_endpoint.clone().unwrap(),
+            )
+            .await
+            .context(OnchainSnafu)?;
 
-        let mut buffer = [0_u8; HASH_SIZE];
-        let bytes = file
-            .take(HASH_SIZE as u64)
-            .read(&mut buffer)
-            .context(ReadHashSnafu { path: path.clone() })?;
-        tracing::trace!("read {bytes} bytes from file");
-        Ok(Hash::new(buffer))
+            match offchain_hash == onchain_hash {
```

Comment:
This will only succeed if the advance runner loads a snapshot that is the first one.

Reply:
Completely overlooked this "detail". Thanks for catching it! 😉